bridge/processor/span.go

```diff
@@ -84,32 +92,133 @@ func (sp *SpanProcessor) checkAndPropose(ctx context.Context) {
 		sp.Logger.Error("Error fetching current child block", "error", err)
 		return
 	}
-	if latestBlock.Number.Uint64() < lastSpan.StartBlock {
+	latestBorBlockNumber := latestBlock.Number.Uint64()
+
+	if latestBorBlockNumber < lastSpan.StartBlock {
 		sp.Logger.Debug("Current bor block is less than last span start block, skipping proposing span", "currentBlock", latestBlock.Number.Uint64(), "lastSpanStartBlock", lastSpan.StartBlock)
 		return
 	}
 
+	var latestMilestoneEndBlock uint64
+	latestMilestone, err := util.GetLatestMilestone(sp.cliCtx.Codec)
+	if err == nil {
+		latestMilestoneEndBlock = latestMilestone.EndBlock
+	}
+
+	// Max of latest milestone end block and latest bor block number
+	// Handle cases where bor is syncing and latest milestone end block is greater than latest bor block number
+	maxBlockNumber := max(latestMilestoneEndBlock, latestBorBlockNumber)
+
 	sp.Logger.Debug("Found last span", "lastSpan", lastSpan.Id, "startBlock", lastSpan.StartBlock, "endBlock", lastSpan.EndBlock)
 
-	nextSpanMsg, err := sp.fetchNextSpanDetails(lastSpan.Id+1, lastSpan.EndBlock+1)
+	if maxBlockNumber > lastSpan.EndBlock {
+		if latestMilestoneEndBlock > lastSpan.EndBlock {
+			sp.Logger.Debug("Bor self commmitted spans, backfill heimdall to fill missing spans", "currentBlock", latestBlock.Number.Uint64(), "lastSpanEndBlock", lastSpan.EndBlock)
+			go func() {
+				defer func() {
+					if r := recover(); r != nil {
+						sp.Logger.Error("Recovered panic in backfillSpans goroutine", "panic", r)
+					}
+				}()
+
+				if err := sp.backfillSpans(ctx, latestMilestoneEndBlock, lastSpan); err != nil {
+					sp.Logger.Error("Error in backfillSpans", "error", err)
+				}
+			}()
+
+		} else {
+			sp.Logger.Debug("Will not backfill heimdall spans, as latest milestone end block is less than last span end block", "currentBlock", latestBlock.Number.Uint64(), "lastSpanEndBlock", lastSpan.EndBlock)
+		}
+	} else {
+		if types.IsBlockCloseToSpanEnd(maxBlockNumber, lastSpan.EndBlock) {
```

Comment:
Hmm. I wonder if this can come to bite us in the normal course when we don't have the next span finalized and are about to reach the end.

Reply:
If this happens, bor will have to self commit and then we backfill.

Reply:
Okay makes sense. Hope we won't run into this that often.